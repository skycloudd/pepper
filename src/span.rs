use core::num::NonZeroUsize;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Span {
    start: usize,
    end: usize,
    file_id: FileId,
}

impl Span {
    pub fn zero(file_id: FileId) -> Self {
        chumsky::span::Span::new(file_id, 0..0)
    }

    pub const fn range(&self) -> core::ops::Range<usize> {
        self.start..self.end
    }
}

impl chumsky::span::Span for Span {
    type Context = FileId;

    type Offset = usize;

    fn new(context: Self::Context, range: core::ops::Range<Self::Offset>) -> Self {
        Self {
            start: range.start,
            end: range.end,
            file_id: context,
        }
    }

    fn context(&self) -> Self::Context {
        self.file_id
    }

    fn start(&self) -> Self::Offset {
        self.start
    }

    fn end(&self) -> Self::Offset {
        self.end
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct FileId(NonZeroUsize);

impl FileId {
    pub fn new(id: usize) -> Option<Self> {
        Some(Self(NonZeroUsize::new(id.checked_add(1)?)?))
    }

    pub fn get(self) -> usize {
        usize::from(self.0) - 1
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Spanned<T>(pub T, Span);

impl<T> Spanned<T> {
    pub const fn new(value: T, span: Span) -> Self {
        Self(value, span)
    }

    pub const fn span(&self) -> Span {
        self.1
    }

    pub fn boxed(self) -> Spanned<Box<T>> {
        Spanned::new(Box::new(self.0), self.1)
    }

    pub fn map<U>(self, f: impl FnOnce(T) -> U) -> Spanned<U> {
        Spanned::new(f(self.0), self.1)
    }

    pub fn map_with_span<U>(self, f: impl FnOnce(T, Span) -> U) -> Spanned<U> {
        Spanned::new(f(self.0, self.1), self.1)
    }

    pub fn map_span(self, f: impl FnOnce(Span) -> Span) -> Self {
        Self(self.0, f(self.1))
    }

    pub const fn as_ref(&self) -> Spanned<&T> {
        Spanned::new(&self.0, self.1)
    }
}

impl<T, E> Spanned<Result<T, E>> {
    pub fn transpose(self) -> Result<Spanned<T>, E> {
        self.0.map(|value| Spanned::new(value, self.1))
    }
}

impl<T> Spanned<Box<T>> {
    pub fn unbox(self) -> Spanned<T> {
        Spanned::new(*self.0, self.1)
    }
}

impl<T: Clone> Spanned<&T> {
    pub fn cloned(self) -> Spanned<T> {
        self.map(Clone::clone)
    }
}

impl<T> core::ops::Deref for Spanned<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T> core::ops::DerefMut for Spanned<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

#[cfg(test)]
impl<T: serde::Serialize> serde::Serialize for Spanned<T> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        self.0.serialize(serializer)
    }
}
