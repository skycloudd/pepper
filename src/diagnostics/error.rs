use super::{Diag, ErrorSpan};
use crate::span::Span;
use chumsky::error::{Rich, RichReason};
use codespan_reporting::diagnostic::Severity;
use core::fmt::Display;
use owo_colors::OwoColorize;
use std::borrow::Cow;

#[derive(Clone, Debug)]
pub enum Error {
    ExpectedFound {
        expected: Vec<String>,
        found: Option<String>,
        span: Span,
    },
    Custom {
        message: String,
        span: Span,
    },
    AmbiguousModule {
        module_name: &'static str,
        module_name_span: Span,
        filenames: Vec<String>,
    },
    MissingModule {
        module_name: &'static str,
        module_name_span: Span,
        filenames: Vec<String>,
    },
}

impl Diag for Error {
    fn message(&self) -> Cow<str> {
        match self {
            Self::ExpectedFound {
                expected,
                found,
                span: _,
            } => if expected.is_empty() {
                "Unexpected end of file".to_string()
            } else if expected.len() == 1 {
                format!(
                    "Expected {}, but found {}",
                    expected[0].green(),
                    found.as_deref().unwrap_or("something else")
                )
            } else {
                format!(
                    "Expected one of {}, but found {}",
                    expected
                        .iter()
                        .map(|expected| expected.green().to_string())
                        .collect::<Vec<_>>()
                        .join(", "),
                    found.as_deref().unwrap_or("something else")
                )
            }
            .into(),
            Self::Custom { message, span: _ } => message.into(),
            Self::AmbiguousModule {
                module_name: _,
                module_name_span: _,
                filenames: _,
            } => "Ambiguous module name".into(),
            Self::MissingModule {
                module_name: _,
                module_name_span: _,
                filenames: _,
            } => "Missing module".into(),
        }
    }

    fn spans(&self) -> Vec<ErrorSpan> {
        match self {
            Self::ExpectedFound {
                expected: _,
                found,
                span,
            } => found.as_ref().map_or_else(
                || vec![ErrorSpan::primary(*span)],
                |found| vec![ErrorSpan::primary_message(format!("Found {found}"), *span)],
            ),
            Self::Custom { message: _, span } => vec![ErrorSpan::primary(*span)],
            Self::AmbiguousModule {
                module_name: _,
                module_name_span,
                filenames: _,
            }
            | Self::MissingModule {
                module_name: _,
                module_name_span,
                filenames: _,
            } => vec![ErrorSpan::primary_message(
                "Declared here",
                *module_name_span,
            )],
        }
    }

    fn notes(&self) -> Vec<String> {
        match self {
            Self::AmbiguousModule {
                module_name,
                module_name_span: _,
                filenames,
            } => core::iter::once(format!(
                "There exist multiple possible files for the module `{module_name}`:"
            ))
            .chain(filenames.iter().map(|filename| format!("  - {filename}")))
            .collect(),
            Self::MissingModule {
                module_name,
                module_name_span: _,
                filenames,
            } => core::iter::once(format!(
                "Could not find a file for the module `{module_name}`"
            ))
            .chain(core::iter::once(
                "These are the possible files for this module:".into(),
            ))
            .chain(filenames.iter().map(|filename| format!("  - {filename}")))
            .collect(),
            Self::ExpectedFound { .. } | Self::Custom { .. } => {
                vec![]
            }
        }
    }

    fn kind(&self) -> Severity {
        Severity::Error
    }
}

pub fn convert(error: &Rich<impl Display, Span, &str>) -> Vec<Error> {
    fn convert_inner(reason: &RichReason<impl Display, &str>, span: Span) -> Vec<Error> {
        match reason {
            RichReason::ExpectedFound { expected, found } => vec![Error::ExpectedFound {
                expected: expected.iter().map(ToString::to_string).collect(),
                found: found.as_ref().map(|f| f.to_string()),
                span,
            }],
            RichReason::Custom(message) => vec![Error::Custom {
                message: message.to_owned(),
                span,
            }],
            RichReason::Many(reasons) => reasons
                .iter()
                .flat_map(|r| convert_inner(r, span))
                .collect(),
        }
    }

    convert_inner(error.reason(), *error.span())
}
