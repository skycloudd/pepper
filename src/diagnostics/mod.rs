use crate::span::Span;
use codespan_reporting::diagnostic::{LabelStyle, Severity};
use std::borrow::Cow;

pub mod error;
pub mod report;

pub trait Diag {
    fn message(&self) -> Cow<str>;
    fn spans(&self) -> Vec<ErrorSpan>;
    fn notes(&self) -> Vec<String>;
    fn kind(&self) -> Severity;
}

#[derive(Debug)]
pub struct ErrorSpan {
    message: Option<String>,
    span: Span,
    label_style: LabelStyle,
}

impl ErrorSpan {
    pub const fn primary(span: Span) -> Self {
        Self {
            message: None,
            span,
            label_style: LabelStyle::Primary,
        }
    }

    pub fn primary_message(message: impl Into<String>, span: Span) -> Self {
        Self {
            message: Some(message.into()),
            span,
            label_style: LabelStyle::Primary,
        }
    }

    pub const fn secondary(span: Span) -> Self {
        Self {
            message: None,
            span,
            label_style: LabelStyle::Secondary,
        }
    }

    pub fn secondary_message(message: impl Into<String>, span: Span) -> Self {
        Self {
            message: Some(message.into()),
            span,
            label_style: LabelStyle::Secondary,
        }
    }
}
