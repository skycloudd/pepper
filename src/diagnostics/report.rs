use super::{Diag, ErrorSpan};
use chumsky::span::Span as _;
use codespan_reporting::diagnostic::{Diagnostic, Label};

pub fn report(diagnostic: &impl Diag) -> Diagnostic<usize> {
    codespan_reporting::diagnostic::Diagnostic::<usize>::new(diagnostic.kind())
        .with_message(diagnostic.message())
        .with_labels(
            diagnostic
                .spans()
                .into_iter()
                .map(|span| match span {
                    ErrorSpan::Primary(message, span) => {
                        let mut label =
                            Label::primary(span.context().id(), span.start()..span.end());

                        if let Some(message) = message {
                            label = label.with_message(message);
                        }

                        label
                    }
                    ErrorSpan::Secondary(message, span) => {
                        let mut label =
                            Label::secondary(span.context().id(), span.start()..span.end());

                        if let Some(message) = message {
                            label = label.with_message(message);
                        }

                        label
                    }
                })
                .collect(),
        )
        .with_notes(diagnostic.notes())
}
