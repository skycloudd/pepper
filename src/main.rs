use camino::Utf8PathBuf;
use chumsky::{input::Input as _, span::Span as _, Parser as _};
use clap::Parser;
use codespan_reporting::{
    files::SimpleFiles,
    term::{
        self,
        termcolor::{ColorChoice, StandardStream},
    },
};
use diagnostics::{error::convert, report::report};
use lasso::ThreadedRodeo;
use span::{FileId, Span};
use std::{fs::read_to_string, process::ExitCode, sync::LazyLock};

mod diagnostics;
mod lexer;
mod parser;
mod scopes;
mod span;
mod typecheck;

static RODEO: LazyLock<ThreadedRodeo> = LazyLock::new(ThreadedRodeo::new);

#[derive(Debug, Parser)]
struct Args {
    filename: Utf8PathBuf,
}

fn main() -> Result<ExitCode, Box<dyn core::error::Error>> {
    let args = Args::parse();

    if !args.filename.is_file() {
        return Err(format!("'{}' is not a file", args.filename).into());
    }

    let mut files = SimpleFiles::new();

    let source = read_to_string(&args.filename)?;

    let file_id = FileId::new(files.add(&args.filename, source.to_string()));

    let mut errors = vec![];

    let (tokens, lexer_errors) = lexer::lexer()
        .parse(source.with_context(file_id))
        .into_output_errors();

    errors.extend(lexer_errors.iter().flat_map(|error| convert(error)));

    let (ast, parser_errors) = tokens.as_ref().map_or_else(
        || (None, vec![]),
        |tokens| {
            let eoi = tokens
                .last()
                .map_or_else(|| Span::zero(file_id), |(_, span)| span.to_end());

            parser::parser()
                .parse(tokens.spanned(eoi))
                .into_output_errors()
        },
    );

    errors.extend(parser_errors.iter().flat_map(|error| convert(error)));

    let (typed_ast, typecheck_errors) = ast.map_or_else(|| (None, vec![]), typecheck::typecheck);

    errors.extend(typecheck_errors);

    if let Some(typed_ast) = typed_ast {
        if cfg!(debug_assertions) {
            eprintln!("{:?}", typed_ast.functions);
        }
    }

    let writer = StandardStream::stderr(ColorChoice::Auto);
    let term_config = term::Config::default();

    for error in &errors {
        let diagnostic = report(error);

        term::emit(&mut writer.lock(), &term_config, &files, &diagnostic)?;
    }

    if errors.is_empty() {
        Ok(ExitCode::SUCCESS)
    } else {
        Ok(ExitCode::FAILURE)
    }
}
