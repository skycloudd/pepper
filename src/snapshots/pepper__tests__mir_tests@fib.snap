---
source: src/main.rs
description: "func main() -> number =\n    fib(10)\n\nfunc fib(n: number) -> number =\n    match n\n    | x where x < 2 => x\n    | _ => fib(n - 1) + fib(n - 2)"
expression: mir
info: tests/mir/fib.pr
snapshot_kind: text
---
functions:
  - name: 0
    params: []
    return_ty:
      Primitive: Number
    body:
      expr:
        Call:
          callee:
            expr:
              Variable: 1
            ty:
              Function:
                params:
                  - Primitive: Number
                return_ty:
                  Primitive: Number
          args:
            - expr:
                Number: 10
              ty:
                Primitive: Number
      ty:
        Primitive: Number
  - name: 1
    params:
      - name: 2
        ty:
          Primitive: Number
    return_ty:
      Primitive: Number
    body:
      expr:
        Match:
          expr:
            expr:
              Variable: 2
            ty:
              Primitive: Number
          arms:
            - pattern:
                pattern_type:
                  Variable: 3
                condition:
                  expr:
                    Intrinsic:
                      LtNumbers:
                        lhs:
                          expr:
                            Variable: 3
                          ty:
                            Primitive: Number
                        rhs:
                          expr:
                            Number: 2
                          ty:
                            Primitive: Number
                  ty:
                    Primitive: Bool
              body:
                expr:
                  Variable: 3
                ty:
                  Primitive: Number
            - pattern:
                pattern_type: Wildcard
                condition: ~
              body:
                expr:
                  Intrinsic:
                    AddNumbers:
                      lhs:
                        expr:
                          Call:
                            callee:
                              expr:
                                Variable: 1
                              ty:
                                Function:
                                  params:
                                    - Primitive: Number
                                  return_ty:
                                    Primitive: Number
                            args:
                              - expr:
                                  Intrinsic:
                                    SubNumbers:
                                      lhs:
                                        expr:
                                          Variable: 2
                                        ty:
                                          Primitive: Number
                                      rhs:
                                        expr:
                                          Number: 1
                                        ty:
                                          Primitive: Number
                                ty:
                                  Primitive: Number
                        ty:
                          Primitive: Number
                      rhs:
                        expr:
                          Call:
                            callee:
                              expr:
                                Variable: 1
                              ty:
                                Function:
                                  params:
                                    - Primitive: Number
                                  return_ty:
                                    Primitive: Number
                            args:
                              - expr:
                                  Intrinsic:
                                    SubNumbers:
                                      lhs:
                                        expr:
                                          Variable: 2
                                        ty:
                                          Primitive: Number
                                      rhs:
                                        expr:
                                          Number: 2
                                        ty:
                                          Primitive: Number
                                ty:
                                  Primitive: Number
                        ty:
                          Primitive: Number
                ty:
                  Primitive: Number
      ty:
        Primitive: Number
